rule/confusing-naming.go

```diff
@@ -0,0 +1,102 @@
+package rule
+
+import (
+	"fmt"
+	"go/ast"
+	"strings"
+
+	"github.com/mgechev/revive/lint"
+)
+
+// ConfusingNamingRule lints method names that differ only by capitalization
+type ConfusingNamingRule struct{}
+
+// Apply applies the rule to given file.
+func (r *ConfusingNamingRule) Apply(file *lint.File, arguments lint.Arguments) []lint.Failure {
+	var failures []lint.Failure
+
+	fileAst := file.AST
+	walker := lintConfusingNames{
+		methodNames: make(map[string][]*ast.Ident),
+		onFailure: func(failure lint.Failure) {
+			failures = append(failures, failure)
+		},
+	}
+
+	ast.Walk(&walker, fileAst)
+
+	return failures
+}
+
+// Name returns the rule name.
+func (r *ConfusingNamingRule) Name() string {
+	return "confusing-naming"
+}
+
+//checkMethodName checks if a given method/function name is similar (just case differences) to other method/function of the same struct/file.
```

Comment:
Maybe struct/package will be better? Are we going to get a warning if:

```go
// a.go
package foo
func bar() {}
```
```go
// b.go
package foo
func baR() {}
```

Reply:
The answer, in short, is: no
The check is _file-wide_. To do it _package-wide_ I can make initialization of `methodNames` struct outside `Apply` (currently, the struct is initialized inside `Apply`, thus for each file)

I have a problem with the _global_ `methodNames` struct: how to differentiate methods/funcs of (actually) different packages? I can prefix package name to methods/funcs but that is not enough. For example many of the projects in the `vendor` directory have a function `main` in a package `main`

Reply:
What do you think about changing:

```diff
// Package represents a package in the project.
type Package struct {
	fset  *token.FileSet
-	files map[string]*File
+	Files map[string]*File

	TypesPkg  *types.Package
	TypesInfo *types.Info

	// sortable is the set of types in the package that implement sort.Interface.
	Sortable map[string]bool
	// main is whether this is a "main" package.
	main int
	mu   sync.Mutex
}
```

This way, from the current file you'll have access to all other files in the package. In `Apply` you can create an index of all functions and the associated files.

The problem is that this will need to be recomputed across all files in the same package. Maybe it's worth exposing a hash called `Symbols` or `Methods` in the package itself:

```
// map[method_name]file_name
type Methods = map[string]string
```

This map can be build before the linter starts analyzing given package. Also, this could be on demand thing, just like type checking.

Let me know what you think about this idea.

Reply:
As you suggest we need a global index of methods of each package. 
I'm drafting a version of the rule that builds the index on-the-fly: in `Apply` we retrieve the list of method names (in fact a `map[string]map[string]bool` as you proposed) of the package corresponding to the file (using `file.Pkg` as key), then we create a `walker` with this list.

You confirm that, in `Apply`, two `lint.File` of the same package have the same `file.Pkg` ?

Reply:
Right [here](https://github.com/mgechev/revive/blob/master/lint/package.go#L151-L162), the package loops over all of its files and lints them one by one.

Reply:
Hi, I've pushed the draft version (it needs refactoring) of the rule. This version detects confusing names of methods/functions of the same package even if they are in different source files. 

:warning: Test does not pass because rule is applied on each file independently. I didn't find a way to test a rule on a set of files.

Reply:
I've pushed a version that, when applied to `revive` source, outputs:

```
vendor/golang.org/x/tools/godoc/vfs/namespace.go:263:21: Method 'Stat' differs only by capitalization to method 'stat' in the same source file
vendor/golang.org/x/tools/godoc/vfs/zipfs/zipfs.go:186:18: Method 'Stat' differs only by capitalization to method 'stat' in the same source file
vendor/golang.org/x/tools/container/intsets/sparse.go:1057:18: Method 'check' differs only by capitalization to method 'Check' in vendor/golang.org/x/tools/container/intsets/util_test.go
vendor/golang.org/x/tools/go/buildutil/allpackages.go:84:6: Method 'allPackages' differs only by capitalization to function 'AllPackages' in the same source file
vendor/golang.org/x/tools/refactor/rename/rename_test.go:1332:6: Method 'main' differs only by capitalization to function 'Main' in vendor/golang.org/x/tools/refactor/rename/rename.go
vendor/github.com/BurntSushi/toml/encode.go:108:21: Method 'encode' differs only by capitalization to method 'Encode' in the same source file
vendor/golang.org/x/tools/go/buildutil/util.go:112:6: Method 'hasSubdir' differs only by capitalization to function 'HasSubdir' in the same source file
vendor/github.com/mattn/go-runewidth/runewidth_posix.go:62:6: Method 'IsEastAsian' differs only by capitalization to function 'isEastAsian' in the same source file
vendor/golang.org/x/tools/go/ast/astutil/imports.go:443:6: Method 'Imports' differs only by capitalization to function 'imports' in the same source file
vendor/golang.org/x/tools/godoc/analysis/analysis.go:267:20: Method 'FileInfo' differs only by capitalization to method 'fileInfo' in the same source file
vendor/github.com/olekukonko/tablewriter/table.go:364:6: Method 'pad' differs only by capitalization to function 'Pad' in vendor/github.com/olekukonko/tablewriter/util.go
vendor/golang.org/x/sys/unix/syscall_linux_amd64.go:78:6: Method 'Gettimeofday' differs only by capitalization to function 'gettimeofday' in vendor/golang.org/x/sys/unix/syscall_linux_amd64_gc.go
vendor/golang.org/x/tools/godoc/index_test.go:137:6: Method 'testIndex' differs only by capitalization to function 'TestIndex' in the same source file
vendor/golang.org/x/tools/godoc/snippet.go:107:24: Method 'NewSnippet' differs only by capitalization to method 'newSnippet' in the same source file
vendor/golang.org/x/tools/godoc/corpus.go:116:6: Method 'NewCorpus' differs only by capitalization to function 'newCorpus' in vendor/golang.org/x/tools/godoc/index_test.go
vendor/golang.org/x/tools/godoc/server.go:741:24: Method 'serveFile' differs only by capitalization to method 'ServeFile' in the same source file
vendor/golang.org/x/tools/godoc/godoc.go:912:24: Method 'WriteNode' differs only by capitalization to method 'writeNode' in the same source file
vendor/golang.org/x/tools/go/ssa/builder.go:2262:19: Method 'build' differs only by capitalization to method 'Build' in the same source file
```